pkg/ingester/ingester.go

```diff
@@ -1643,6 +1648,168 @@ func (i *Ingester) queryStreamChunks(ctx context.Context, db *userTSDB, from, th
 	return numSeries, numSamples, nil
 }
 
+func (i *Ingester) queryStreamStreaming(ctx context.Context, db *userTSDB, from, through int64, matchers []*labels.Matcher, shard *sharding.ShardSelector, stream client.Ingester_QueryStreamServer, batchSize uint64) (numSeries, numSamples int, _ error) {
+	allIterators, err := i.sendStreamingQuerySeries(ctx, db, from, through, matchers, shard, stream)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	numSamples, err = i.sendStreamingQueryChunks(allIterators, stream, batchSize)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return len(allIterators), numSamples, nil
+}
+
+func (i *Ingester) sendStreamingQuerySeries(ctx context.Context, db *userTSDB, from, through int64, matchers []*labels.Matcher, shard *sharding.ShardSelector, stream client.Ingester_QueryStreamServer) ([]chunks.Iterator, error) {
+	var q storage.ChunkQuerier
+	var err error
+	if i.limits.OutOfOrderTimeWindow(db.userID) > 0 {
+		q, err = db.UnorderedChunkQuerier(ctx, from, through)
+	} else {
+		q, err = db.ChunkQuerier(ctx, from, through)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: does q need to remain open until we're done using the series and their chunks?
```

Comment:
The answer is yes. Else the query results could be wrong or ingester can error out if head compaction happens between the two calls (one for series and another for chunks)

Reply:
Thanks for clarifying, I've fixed this in 230be51cf24855fc4e862425e1042da2f116ad59